Userland/Libraries/LibSoftGPU/Sampler.cpp

```diff
@@ -111,9 +111,44 @@ Vector4<AK::SIMD::f32x4> Sampler::sample_2d(Vector2<AK::SIMD::f32x4> const& uv)
 
     auto const& image = *m_config.bound_image;
 
+    constexpr unsigned base_level = 0; // FIXME: make base level configurable with glTexParameteri(GL_TEXTURE_BASE_LEVEL, base_level)
+
+    // determine texture scale factor. OpenGL 1.5 spec chapter 3.8.8
+    auto texel_coordinates = uv;
+    texel_coordinates.set_x(texel_coordinates.x() * static_cast<float>(image.level_width(base_level)));
+    texel_coordinates.set_y(texel_coordinates.y() * static_cast<float>(image.level_height(base_level)));
```

Comment:
These `static_cast`s are hiding possible truncation. It probably does not make a lot of sense that `level_width/height/depth` return an `unsigned`? An `u8` is probably enough and removes the need for explicit casting here.

Reply:
> It probably does not make a lot of sense that level_width/height/depth return an unsigned?

I don't think there's any guarantee that a mip level with be maximally bound by a `u8`??? What happens if we have a texture that's 2048x2048? The first mip will be `1024x1024`?

Reply:
Ah I misinterpreted the functions, I figured they'd return a level. Some sane upper limit could still be applied though, i.e. `u16` allows for 65536x65536 which is about 17 gigabytes of RGBA data for a single level.

Reply:
I guess we could make it an unsigned short? 65536px should be big enough

Reply:
Agreed, I can't imagine a texture being larger than that realistically. Some id games might break this assumption if they use mega-texturing, however.

Reply:
Ok I'll look into it tomorrow. This will require lots of changes to the image class.

Reply:
Leaving it as-is and adding a couple of `FIXME`s to `Image` would be fine with me!